### PR TITLE
Support jsonl responses from llms

### LINKED ISSUE #20: Response parsing should accommodate JSONL responses
Ollama (and other) LLMs return JSONL responses. For example:

```json
{"model":"qwen2.5-coder:32b","created_at":"2024-12-04T01:02:49.430146Z","response":"{\"","done":false}
{"model":"qwen2.5-coder:32b","created_at":"2024-12-04T01:02:49.547325Z","response":"tool","done":false}
<snip>
"model":"qwen2.5-coder:32b","created_at":"2024-12-04T01:02:51.164084Z","response":"","done":true,"done_reason":"stop","context":[<snip>],"total_duration":4234087125,"load_duration":9450125,"prompt_eval_count":279,"prompt_eval_duration":2489000000,"eval_count":16,"eval_duration":1734000000}
```

The `attemptGenerate` function of `LLMImpl` fails to handle this gracefully, resulting in warnings for every invocation to Ollama.

### PR DESCRIPTION
**Pull request by matthewmcneely**
This PR gracefully handles processing of JSONL responses from LLMs. Closes #20.

See comments in PR for other changes needed to get testing to work.

### CODE REVIEW

File: gollm_test.go [hunk 1]
```diff
@@ -1,3 +1,8 @@
+//go:build ignore
```
Comment (matthewmcneely) [inline gollm_test.go:1]: The `WithProviderRegistry` function is not present in the package. Maybe it's on another branch? I had to comment this out in order to get tests to work.

File: live_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLiveAPICallWithGenerate(t *testing.T) {
```
Comment (matthewmcneely) [inline live_test.go:15]: I added this "live" test to verify that I hadn't broken openai response parsing. I added an ollama test as well. I don't have keys to other providers, but those that do could add similar Runners.

Note, the tests are skipped if the API keys (for remote services like openai) or valid endpoints (like Ollama) are not present.

File: llm/errors_test.go [hunk 3]
```diff
@@ -5,6 +5,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
```
Comment (matthewmcneely) [inline llm/errors_test.go:8]: Not sure how tests were passing without this import.

File: llm/llm.go [hunk 4]
```diff
@@ -248,19 +248,27 @@ func (l *LLMImpl) attemptGenerate(ctx context.Context, prompt *Prompt) (string,
 
 	// Extract and log caching information
 	var fullResponse map[string]interface{}
-	if err := json.Unmarshal(body, &fullResponse); err == nil {
-		if usage, ok := fullResponse["usage"].(map[string]interface{}); ok {
-			l.logger.Debug("Usage information", "usage", usage)
-			cacheInfo := map[string]interface{}{
-				"cache_creation_input_tokens": usage["cache_creation_input_tokens"],
-				"cache_read_input_tokens":     usage["cache_read_input_tokens"],
+	if err := json.Unmarshal(body, &fullResponse); err != nil {
```
Comment (matthewmcneely) [inline llm/llm.go:251]: Here's the meat of the PR. If JSON parsing fails, we try parsing JSONL and getting the last entry.

**Comment by teilomillet** [comment]
Sorry for the delay, thank you @matthewmcneely !!

**Pull request merged**